File: bindings/haskell/src/ProvenServers/Kerberos.hs
-- SPDX-License-Identifier: PMPL-1.0-or-later
-- Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
--
-- | Kerberos protocol types for the proven-servers ABI.
--
-- All tag values match the Idris2 ABI discriminants exactly.

module ProvenServers.Kerberos
  (
    kerberosPort
  , MessageType(..)
  , messageTypeToTag
  , messageTypeFromTag
  , isRequest
  , isReply
  , EncryptionType(..)
  , encryptionTypeToTag
  , encryptionTypeFromTag
  , isLegacy
  , PrincipalType(..)
  , principalTypeToTag
  , principalTypeFromTag
  , TicketFlag(..)
  , ticketFlagToTag
  , ticketFlagFromTag
  , isDelegation
  , ErrorCode(..)
  , errorCodeToTag
  , errorCodeFromTag
  , isSuccess
  , AuthState(..)
  , authStateToTag
  , authStateFromTag
  , authStateCanTransitionTo
  , EncStrength(..)
  , encStrengthToTag
  , encStrengthFromTag
  , PreAuthType(..)
  , preAuthTypeToTag
  , preAuthTypeFromTag
  , NegotiationState(..)
  , negotiationStateToTag
  , negotiationStateFromTag
  ) where

import Data.Word (Word16, Word8)

-- | Standard Kerberos KDC port (RFC 4120).
kerberosPort :: Word16
kerberosPort = 88

-- ---------------------------------------------------------------------------
-- MessageType
-- ---------------------------------------------------------------------------

-- | Standard Kerberos KDC port (RFC 4120).
--
-- Tags 0-9 (10 constructors).
data MessageType
  = AsReq  -- ^ AS-REQ — Authentication Service request (tag 0).
  | AsRep  -- ^ AS-REP — Authentication Service reply (tag 1).
  | TgsReq  -- ^ TGS-REQ — Ticket-Granting Service request (tag 2).
  | TgsRep  -- ^ TGS-REP — Ticket-Granting Service reply (tag 3).
  | ApReq  -- ^ AP-REQ — Application request (tag 4).
  | ApRep  -- ^ AP-REP — Application reply (tag 5).
  | KrbError  -- ^ KRB-ERROR — Error message (tag 6).
  | KrbSafe  -- ^ KRB-SAFE — Safe (authenticated) message (tag 7).
  | KrbPriv  -- ^ KRB-PRIV — Private (encrypted) message (tag 8).
  | KrbCred  -- ^ KRB-CRED — Credential forwarding (tag 9).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'MessageType' to its ABI tag value.
messageTypeToTag :: MessageType -> Word8
messageTypeToTag = fromIntegral . fromEnum

-- | Decode a 'MessageType' from its ABI tag value.
messageTypeFromTag :: Word8 -> Maybe MessageType
messageTypeFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: MessageType)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- | Whether this message is a request.
isRequest :: MessageType -> Bool
isRequest AsReq = True
isRequest TgsReq = True
isRequest ApReq = True
isRequest _ = False

-- | Whether this message is a reply.
isReply :: MessageType -> Bool
isReply AsRep = True
isReply TgsRep = True
isReply ApRep = True
isReply _ = False

-- ---------------------------------------------------------------------------
-- EncryptionType
-- ---------------------------------------------------------------------------

-- | Kerberos encryption types (RFC 3961).
--
-- Tags 0-4 (5 constructors).
data EncryptionType
  = Aes256CtsHmacSha1  -- ^ AES256-CTS-HMAC-SHA1-96 (tag 0).
  | Aes128CtsHmacSha1  -- ^ AES128-CTS-HMAC-SHA1-96 (tag 1).
  | Aes256CtsHmacSha384  -- ^ AES256-CTS-HMAC-SHA384-192 (tag 2).
  | Rc4Hmac  -- ^ RC4-HMAC (legacy, tag 3).
  | Des3CbcSha1  -- ^ DES3-CBC-SHA1 (legacy, tag 4).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'EncryptionType' to its ABI tag value.
encryptionTypeToTag :: EncryptionType -> Word8
encryptionTypeToTag = fromIntegral . fromEnum

-- | Decode a 'EncryptionType' from its ABI tag value.
encryptionTypeFromTag :: Word8 -> Maybe EncryptionType
encryptionTypeFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: EncryptionType)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- | Whether this encryption type is considered legacy/deprecated.
isLegacy :: EncryptionType -> Bool
isLegacy Rc4Hmac = True
isLegacy Des3CbcSha1 = True
isLegacy _ = False

-- ---------------------------------------------------------------------------
-- PrincipalType
-- ---------------------------------------------------------------------------

-- | Kerberos principal name types (RFC 4120).
--
-- Tags 0-6 (7 constructors).
data PrincipalType
  = NtUnknown  -- ^ NT-UNKNOWN (tag 0).
  | NtPrincipal  -- ^ NT-PRINCIPAL — general principal (tag 1).
  | NtSrvInst  -- ^ NT-SRV-INST — service instance (tag 2).
  | NtSrvHst  -- ^ NT-SRV-HST — service with host (tag 3).
  | NtUid  -- ^ NT-UID — unique ID (tag 4).
  | NtX500  -- ^ NT-X500-PRINCIPAL — X.500 principal (tag 5).
  | NtEnterprise  -- ^ NT-ENTERPRISE — enterprise principal (tag 6).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'PrincipalType' to its ABI tag value.
principalTypeToTag :: PrincipalType -> Word8
principalTypeToTag = fromIntegral . fromEnum

-- | Decode a 'PrincipalType' from its ABI tag value.
principalTypeFromTag :: Word8 -> Maybe PrincipalType
principalTypeFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: PrincipalType)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- ---------------------------------------------------------------------------
-- TicketFlag
-- ---------------------------------------------------------------------------

-- | Kerberos ticket flags (RFC 4120).
--
-- Tags 0-6 (7 constructors).
data TicketFlag
  = Forwardable  -- ^ Ticket may be forwarded (tag 0).
  | Forwarded  -- ^ Ticket has been forwarded (tag 1).
  | Proxiable  -- ^ Ticket may be proxied (tag 2).
  | Proxy  -- ^ Ticket is a proxy (tag 3).
  | Renewable  -- ^ Ticket may be renewed (tag 4).
  | PreAuthent  -- ^ Client was pre-authenticated (tag 5).
  | HwAuthent  -- ^ Hardware authentication was used (tag 6).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'TicketFlag' to its ABI tag value.
ticketFlagToTag :: TicketFlag -> Word8
ticketFlagToTag = fromIntegral . fromEnum

-- | Decode a 'TicketFlag' from its ABI tag value.
ticketFlagFromTag :: Word8 -> Maybe TicketFlag
ticketFlagFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: TicketFlag)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- | Whether this flag relates to delegation.
isDelegation :: TicketFlag -> Bool
isDelegation Forwardable = True
isDelegation Forwarded = True
isDelegation Proxiable = True
isDelegation Proxy = True
isDelegation _ = False

-- ---------------------------------------------------------------------------
-- ErrorCode
-- ---------------------------------------------------------------------------

-- | Kerberos KDC error codes (RFC 4120).
--
-- Tags 0-9 (10 constructors).
data ErrorCode
  = KdcErrNone  -- ^ KDC_ERR_NONE — no error (tag 0).
  | KdcErrNameExp  -- ^ KDC_ERR_NAME_EXP — client name expired (tag 1).
  | KdcErrServiceExp  -- ^ KDC_ERR_SERVICE_EXP — service name expired (tag 2).
  | KdcErrBadPvno  -- ^ KDC_ERR_BAD_PVNO — bad protocol version (tag 3).
  | KdcErrCOldMastKvno  -- ^ KDC_ERR_C_OLD_MAST_KVNO — client key version too old (tag 4).
  | KdcErrSOldMastKvno  -- ^ KDC_ERR_S_OLD_MAST_KVNO — server key version too old (tag 5).
  | KdcErrCPrincipalUnknown  -- ^ KDC_ERR_C_PRINCIPAL_UNKNOWN — client principal not found (tag 6).
  | KdcErrSPrincipalUnknown  -- ^ KDC_ERR_S_PRINCIPAL_UNKNOWN — service principal not found (tag 7).
  | KdcErrPreauthFailed  -- ^ KDC_ERR_PREAUTH_FAILED — pre-authentication failed (tag 8).
  | KdcErrPreauthRequired  -- ^ KDC_ERR_PREAUTH_REQUIRED — pre-authentication required (tag 9).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'ErrorCode' to its ABI tag value.
errorCodeToTag :: ErrorCode -> Word8
errorCodeToTag = fromIntegral . fromEnum

-- | Decode a 'ErrorCode' from its ABI tag value.
errorCodeFromTag :: Word8 -> Maybe ErrorCode
errorCodeFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: ErrorCode)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- | Whether this code indicates success.
isSuccess :: ErrorCode -> Bool
isSuccess KdcErrNone = True
isSuccess _ = False

-- ---------------------------------------------------------------------------
-- AuthState
-- ---------------------------------------------------------------------------

-- | Kerberos authentication state machine.
--
-- Tags 0-4 (5 constructors).
data AuthState
  = Initial  -- ^ Initial — no tickets (tag 0).
  | TgtObtained  -- ^ TGT obtained from AS (tag 1).
  | ServiceTicketObtained  -- ^ Service ticket obtained from TGS (tag 2).
  | Authenticated  -- ^ Authenticated — AP-REP received (tag 3).
  | AuthFailed  -- ^ Authentication failed (tag 4).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'AuthState' to its ABI tag value.
authStateToTag :: AuthState -> Word8
authStateToTag = fromIntegral . fromEnum

-- | Decode a 'AuthState' from its ABI tag value.
authStateFromTag :: Word8 -> Maybe AuthState
authStateFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: AuthState)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- | Validate whether a state transition is allowed.
authStateCanTransitionTo :: AuthState -> AuthState -> Bool
authStateCanTransitionTo Initial TgtObtained = True
authStateCanTransitionTo TgtObtained ServiceTicketObtained = True
authStateCanTransitionTo ServiceTicketObtained Authenticated = True
authStateCanTransitionTo Initial AuthFailed = True
authStateCanTransitionTo TgtObtained AuthFailed = True
authStateCanTransitionTo ServiceTicketObtained AuthFailed = True
authStateCanTransitionTo _ _ = False

-- ---------------------------------------------------------------------------
-- EncStrength
-- ---------------------------------------------------------------------------

-- | Encryption strength classification.
--
-- Tags 0-2 (3 constructors).
data EncStrength
  = Strong  -- ^ Strong — recommended (tag 0).
  | Medium  -- ^ Medium — acceptable (tag 1).
  | Weak  -- ^ Weak — deprecated (tag 2).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'EncStrength' to its ABI tag value.
encStrengthToTag :: EncStrength -> Word8
encStrengthToTag = fromIntegral . fromEnum

-- | Decode a 'EncStrength' from its ABI tag value.
encStrengthFromTag :: Word8 -> Maybe EncStrength
encStrengthFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: EncStrength)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- ---------------------------------------------------------------------------
-- PreAuthType
-- ---------------------------------------------------------------------------

-- | Kerberos pre-authentication types.
--
-- Tags 0-3 (4 constructors).
data PreAuthType
  = PaEncTimestamp  -- ^ PA-ENC-TIMESTAMP — encrypted timestamp (tag 0).
  | PaEtypeInfo2  -- ^ PA-ETYPE-INFO2 — encryption type info (tag 1).
  | PaFxFast  -- ^ PA-FX-FAST — Flexible Authentication (tag 2).
  | PaFxCookie  -- ^ PA-FX-COOKIE — FAST cookie (tag 3).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'PreAuthType' to its ABI tag value.
preAuthTypeToTag :: PreAuthType -> Word8
preAuthTypeToTag = fromIntegral . fromEnum

-- | Decode a 'PreAuthType' from its ABI tag value.
preAuthTypeFromTag :: Word8 -> Maybe PreAuthType
preAuthTypeFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: PreAuthType)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- ---------------------------------------------------------------------------
-- NegotiationState
-- ---------------------------------------------------------------------------

-- | Kerberos encryption negotiation state.
--
-- Tags 0-3 (4 constructors).
data NegotiationState
  = NegIdle  -- ^ No negotiation started (tag 0).
  | Proposed  -- ^ Client proposed encryption types (tag 1).
  | Selected  -- ^ Server selected an encryption type (tag 2).
  | NegFailed  -- ^ Negotiation failed — no common type (tag 3).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'NegotiationState' to its ABI tag value.
negotiationStateToTag :: NegotiationState -> Word8
negotiationStateToTag = fromIntegral . fromEnum

-- | Decode a 'NegotiationState' from its ABI tag value.
negotiationStateFromTag :: Word8 -> Maybe NegotiationState
negotiationStateFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: NegotiationState)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

File: bindings/elixir/lib/proven_servers/kerberos.ex
# SPDX-License-Identifier: MPL-2.0
# Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
defmodule ProvenServers.Kerberos do
  @moduledoc """
  Kerberos protocol types for the proven-servers ABI.
  
  Mirrors the Idris2 module `KerberosABI.Types` and its type definitions:
  - `MessageType`       — Kerberos message types (10 constructors, tags 0-9)
  - `EncryptionType`    — Encryption algorithms (5 constructors, tags 0-4)
  - `PrincipalType`     — Principal name types (7 constructors, tags 0-6)
  - `TicketFlag`        — Ticket flags (7 constructors, tags 0-6)
  - `ErrorCode`         — KDC error codes (10 constructors, tags 0-9)
  - `AuthState`         — Authentication state machine (5 constructors, tags 0-4)
  - `EncStrength`       — Encryption strength levels (3 constructors, tags 0-2)
  - `PreAuthType`       — Pre-authentication types (4 constructors, tags 0-3)
  - `NegotiationState`  — Negotiation state machine (4 constructors, tags 0-3)
  
  All discriminant values match the Idris2 ABI tag definitions exactly.

  All tag values match the Idris2 ABI definitions exactly.
  """

  @doc "Standard Kerberos KDC port (RFC 4120)."
  @spec kerberos_port() :: non_neg_integer()
  def kerberos_port, do: 88

  # ===========================================================================
  # MessageType (tags 0-9)
  # ===========================================================================

  @typedoc """
  MessageType types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type message_type ::
          :as_req
          | :as_rep
          | :tgs_req
          | :tgs_rep
          | :ap_req
          | :ap_rep
          | :krb_error
          | :krb_safe
          | :krb_priv
          | :krb_cred

  @message_type_tags %{
    as_req: 0,
    as_rep: 1,
    tgs_req: 2,
    tgs_rep: 3,
    ap_req: 4,
    ap_rep: 5,
    krb_error: 6,
    krb_safe: 7,
    krb_priv: 8,
    krb_cred: 9,
  }

  @tag_to_message_type Map.new(@message_type_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `MessageType` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..9, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Kerberos.message_type_from_tag(0)
      {:ok, :as_req}
  """
  @spec message_type_from_tag(non_neg_integer()) :: {:ok, message_type()} | :error
  def message_type_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 9 do
    {:ok, Map.fetch!(@tag_to_message_type, tag)}
  end

  def message_type_from_tag(_tag), do: :error

  @doc """
  Encode a `MessageType` to the C-ABI tag value.
  """
  @spec message_type_to_tag(message_type()) :: non_neg_integer()
  def message_type_to_tag(val) when is_map_key(@message_type_tags, val) do
    Map.fetch!(@message_type_tags, val)
  end

  @doc """
  All `MessageType` variants in tag order.
  """
  @spec all_message_types() :: [message_type()]
  def all_message_types do
    [
      :as_req, :as_rep, :tgs_req, :tgs_rep, :ap_req, :ap_rep, :krb_error,
      :krb_safe, :krb_priv, :krb_cred
    ]
  end

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  Whether this message is a request.
  """
  @spec is_request?(message_type()) :: boolean()
  def is_request?(val) when val in [:as_req, :tgs_req, :ap_req], do: true
  def is_request?(_val), do: false

  @doc """
  Whether this message is a reply.
  """
  @spec is_reply?(message_type()) :: boolean()
  def is_reply?(val) when val in [:as_rep, :tgs_rep, :ap_rep], do: true
  def is_reply?(_val), do: false

  # ===========================================================================
  # EncryptionType (tags 0-4)
  # ===========================================================================

  @typedoc """
  EncryptionType types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type encryption_type ::
          :aes256_cts_hmac_sha1
          | :aes128_cts_hmac_sha1
          | :aes256_cts_hmac_sha384
          | :rc4_hmac
          | :des3_cbc_sha1

  @encryption_type_tags %{
    aes256_cts_hmac_sha1: 0,
    aes128_cts_hmac_sha1: 1,
    aes256_cts_hmac_sha384: 2,
    rc4_hmac: 3,
    des3_cbc_sha1: 4,
  }

  @tag_to_encryption_type Map.new(@encryption_type_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `EncryptionType` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..4, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Kerberos.encryption_type_from_tag(0)
      {:ok, :aes256_cts_hmac_sha1}
  """
  @spec encryption_type_from_tag(non_neg_integer()) :: {:ok, encryption_type()} | :error
  def encryption_type_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 4 do
    {:ok, Map.fetch!(@tag_to_encryption_type, tag)}
  end

  def encryption_type_from_tag(_tag), do: :error

  @doc """
  Encode a `EncryptionType` to the C-ABI tag value.
  """
  @spec encryption_type_to_tag(encryption_type()) :: non_neg_integer()
  def encryption_type_to_tag(val) when is_map_key(@encryption_type_tags, val) do
    Map.fetch!(@encryption_type_tags, val)
  end

  @doc """
  All `EncryptionType` variants in tag order.
  """
  @spec all_encryption_types() :: [encryption_type()]
  def all_encryption_types do
    [
      :aes256_cts_hmac_sha1, :aes128_cts_hmac_sha1, :aes256_cts_hmac_sha384,
      :rc4_hmac, :des3_cbc_sha1
    ]
  end

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  The encryption strength classification.
        match self {

  Whether this encryption type is considered legacy/deprecated.
  """
  @spec is_legacy?(encryption_type()) :: boolean()
  def is_legacy?(val) when val in [:rc4_hmac, :des3_cbc_sha1], do: true
  def is_legacy?(_val), do: false

  # ===========================================================================
  # PrincipalType (tags 0-6)
  # ===========================================================================

  @typedoc """
  PrincipalType types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type principal_type ::
          :nt_unknown
          | :nt_principal
          | :nt_srv_inst
          | :nt_srv_hst
          | :nt_uid
          | :nt_x500
          | :nt_enterprise

  @principal_type_tags %{
    nt_unknown: 0,
    nt_principal: 1,
    nt_srv_inst: 2,
    nt_srv_hst: 3,
    nt_uid: 4,
    nt_x500: 5,
    nt_enterprise: 6,
  }

  @tag_to_principal_type Map.new(@principal_type_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `PrincipalType` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..6, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Kerberos.principal_type_from_tag(0)
      {:ok, :nt_unknown}
  """
  @spec principal_type_from_tag(non_neg_integer()) :: {:ok, principal_type()} | :error
  def principal_type_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 6 do
    {:ok, Map.fetch!(@tag_to_principal_type, tag)}
  end

  def principal_type_from_tag(_tag), do: :error

  @doc """
  Encode a `PrincipalType` to the C-ABI tag value.
  """
  @spec principal_type_to_tag(principal_type()) :: non_neg_integer()
  def principal_type_to_tag(val) when is_map_key(@principal_type_tags, val) do
    Map.fetch!(@principal_type_tags, val)
  end

  @doc """
  All `PrincipalType` variants in tag order.
  """
  @spec all_principal_types() :: [principal_type()]
  def all_principal_types do
    [
      :nt_unknown, :nt_principal, :nt_srv_inst, :nt_srv_hst, :nt_uid,
      :nt_x500, :nt_enterprise
    ]
  end

  # ===========================================================================
  # TicketFlag (tags 0-6)
  # ===========================================================================

  @typedoc """
  TicketFlag types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type ticket_flag ::
          :forwardable
          | :forwarded
          | :proxiable
          | :proxy
          | :renewable
          | :pre_authent
          | :hw_authent

  @ticket_flag_tags %{
    forwardable: 0,
    forwarded: 1,
    proxiable: 2,
    proxy: 3,
    renewable: 4,
    pre_authent: 5,
    hw_authent: 6,
  }

  @tag_to_ticket_flag Map.new(@ticket_flag_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `TicketFlag` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..6, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Kerberos.ticket_flag_from_tag(0)
      {:ok, :forwardable}
  """
  @spec ticket_flag_from_tag(non_neg_integer()) :: {:ok, ticket_flag()} | :error
  def ticket_flag_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 6 do
    {:ok, Map.fetch!(@tag_to_ticket_flag, tag)}
  end

  def ticket_flag_from_tag(_tag), do: :error

  @doc """
  Encode a `TicketFlag` to the C-ABI tag value.
  """
  @spec ticket_flag_to_tag(ticket_flag()) :: non_neg_integer()
  def ticket_flag_to_tag(val) when is_map_key(@ticket_flag_tags, val) do
    Map.fetch!(@ticket_flag_tags, val)
  end

  @doc """
  All `TicketFlag` variants in tag order.
  """
  @spec all_ticket_flags() :: [ticket_flag()]
  def all_ticket_flags do
    [
      :forwardable, :forwarded, :proxiable, :proxy, :renewable, :pre_authent,
      :hw_authent
    ]
  end

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  Whether this flag relates to delegation.
  """
  @spec is_delegation?(ticket_flag()) :: boolean()
  def is_delegation?(val) when val in [:forwardable, :forwarded, :proxiable, :proxy], do: true
  def is_delegation?(_val), do: false

  # ===========================================================================
  # ErrorCode (tags 0-9)
  # ===========================================================================

  @typedoc """
  ErrorCode types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type error_code ::
          :kdc_err_none
          | :kdc_err_name_exp
          | :kdc_err_service_exp
          | :kdc_err_bad_pvno
          | :kdc_err_c_old_mast_kvno
          | :kdc_err_s_old_mast_kvno
          | :kdc_err_c_principal_unknown
          | :kdc_err_s_principal_unknown
          | :kdc_err_preauth_failed
          | :kdc_err_preauth_required

  @error_code_tags %{
    kdc_err_none: 0,
    kdc_err_name_exp: 1,
    kdc_err_service_exp: 2,
    kdc_err_bad_pvno: 3,
    kdc_err_c_old_mast_kvno: 4,
    kdc_err_s_old_mast_kvno: 5,
    kdc_err_c_principal_unknown: 6,
    kdc_err_s_principal_unknown: 7,
    kdc_err_preauth_failed: 8,
    kdc_err_preauth_required: 9,
  }

  @tag_to_error_code Map.new(@error_code_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `ErrorCode` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..9, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Kerberos.error_code_from_tag(0)
      {:ok, :kdc_err_none}
  """
  @spec error_code_from_tag(non_neg_integer()) :: {:ok, error_code()} | :error
  def error_code_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 9 do
    {:ok, Map.fetch!(@tag_to_error_code, tag)}
  end

  def error_code_from_tag(_tag), do: :error

  @doc """
  Encode a `ErrorCode` to the C-ABI tag value.
  """
  @spec error_code_to_tag(error_code()) :: non_neg_integer()
  def error_code_to_tag(val) when is_map_key(@error_code_tags, val) do
    Map.fetch!(@error_code_tags, val)
  end

  @doc """
  All `ErrorCode` variants in tag order.
  """
  @spec all_error_codes() :: [error_code()]
  def all_error_codes do
    [
      :kdc_err_none, :kdc_err_name_exp, :kdc_err_service_exp, :kdc_err_bad_pvno,
      :kdc_err_c_old_mast_kvno, :kdc_err_s_old_mast_kvno, :kdc_err_c_principal_unknown,
      :kdc_err_s_principal_unknown, :kdc_err_preauth_failed, :kdc_err_preauth_required,
    ]
  end

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  Whether this code indicates success.
  """
  @spec is_success?(error_code()) :: boolean()
  def is_success?(val) when val in [:kdc_err_none], do: true
  def is_success?(_val), do: false

  # ===========================================================================
  # AuthState (tags 0-4)
  # ===========================================================================

  @typedoc """
  AuthState types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type auth_state ::
          :initial
          | :tgt_obtained
          | :service_ticket_obtained
          | :authenticated
          | :auth_failed

  @auth_state_tags %{
    initial: 0,
    tgt_obtained: 1,
    service_ticket_obtained: 2,
    authenticated: 3,
    auth_failed: 4,
  }

  @tag_to_auth_state Map.new(@auth_state_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `AuthState` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..4, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Kerberos.auth_state_from_tag(0)
      {:ok, :initial}
  """
  @spec auth_state_from_tag(non_neg_integer()) :: {:ok, auth_state()} | :error
  def auth_state_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 4 do
    {:ok, Map.fetch!(@tag_to_auth_state, tag)}
  end

  def auth_state_from_tag(_tag), do: :error

  @doc """
  Encode a `AuthState` to the C-ABI tag value.
  """
  @spec auth_state_to_tag(auth_state()) :: non_neg_integer()
  def auth_state_to_tag(val) when is_map_key(@auth_state_tags, val) do
    Map.fetch!(@auth_state_tags, val)
  end

  @doc """
  All `AuthState` variants in tag order.
  """
  @spec all_auth_states() :: [auth_state()]
  def all_auth_states do
    [
      :initial, :tgt_obtained, :service_ticket_obtained, :authenticated,
      :auth_failed
    ]
  end

  # validate_auth_state_transition removed: unproven reimplementation. The verified check lives in the
  # Idris2/Zig core; calling it needs FFI wiring not yet present in this binding.
  # Do not reimplement here. See docs/decisions/0003-keep-bindings-thin-abi-wrappers.md

  # ===========================================================================
  # EncStrength (tags 0-2)
  # ===========================================================================

  @typedoc """
  EncStrength types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type enc_strength :: :strong | :medium | :weak

  @enc_strength_tags %{
    strong: 0,
    medium: 1,
    weak: 2,
  }

  @tag_to_enc_strength Map.new(@enc_strength_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `EncStrength` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..2, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Kerberos.enc_strength_from_tag(0)
      {:ok, :strong}
  """
  @spec enc_strength_from_tag(non_neg_integer()) :: {:ok, enc_strength()} | :error
  def enc_strength_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 2 do
    {:ok, Map.fetch!(@tag_to_enc_strength, tag)}
  end

  def enc_strength_from_tag(_tag), do: :error

  @doc """
  Encode a `EncStrength` to the C-ABI tag value.
  """
  @spec enc_strength_to_tag(enc_strength()) :: non_neg_integer()
  def enc_strength_to_tag(val) when is_map_key(@enc_strength_tags, val) do
    Map.fetch!(@enc_strength_tags, val)
  end

  @doc """
  All `EncStrength` variants in tag order.
  """
  @spec all_enc_strengths() :: [enc_strength()]
  def all_enc_strengths, do: [:strong, :medium, :weak]

  # ===========================================================================
  # PreAuthType (tags 0-3)
  # ===========================================================================

  @typedoc """
  PreAuthType types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type pre_auth_type :: :pa_enc_timestamp | :pa_etype_info2 | :pa_fx_fast | :pa_fx_cookie

  @pre_auth_type_tags %{
    pa_enc_timestamp: 0,
    pa_etype_info2: 1,
    pa_fx_fast: 2,
    pa_fx_cookie: 3,
  }

  @tag_to_pre_auth_type Map.new(@pre_auth_type_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `PreAuthType` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..3, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Kerberos.pre_auth_type_from_tag(0)
      {:ok, :pa_enc_timestamp}
  """
  @spec pre_auth_type_from_tag(non_neg_integer()) :: {:ok, pre_auth_type()} | :error
  def pre_auth_type_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 3 do
    {:ok, Map.fetch!(@tag_to_pre_auth_type, tag)}
  end

  def pre_auth_type_from_tag(_tag), do: :error

  @doc """
  Encode a `PreAuthType` to the C-ABI tag value.
  """
  @spec pre_auth_type_to_tag(pre_auth_type()) :: non_neg_integer()
  def pre_auth_type_to_tag(val) when is_map_key(@pre_auth_type_tags, val) do
    Map.fetch!(@pre_auth_type_tags, val)
  end

  @doc """
  All `PreAuthType` variants in tag order.
  """
  @spec all_pre_auth_types() :: [pre_auth_type()]
  def all_pre_auth_types, do: [:pa_enc_timestamp, :pa_etype_info2, :pa_fx_fast, :pa_fx_cookie]

  # ===========================================================================
  # NegotiationState (tags 0-3)
  # ===========================================================================

  @typedoc """
  NegotiationState types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type negotiation_state :: :neg_idle | :proposed | :selected | :neg_failed

  @negotiation_state_tags %{
    neg_idle: 0,
    proposed: 1,
    selected: 2,
    neg_failed: 3,
  }

  @tag_to_negotiation_state Map.new(@negotiation_state_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `NegotiationState` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..3, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Kerberos.negotiation_state_from_tag(0)
      {:ok, :neg_idle}
  """
  @spec negotiation_state_from_tag(non_neg_integer()) :: {:ok, negotiation_state()} | :error
  def negotiation_state_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 3 do
    {:ok, Map.fetch!(@tag_to_negotiation_state, tag)}
  end

  def negotiation_state_from_tag(_tag), do: :error

  @doc """
  Encode a `NegotiationState` to the C-ABI tag value.
  """
  @spec negotiation_state_to_tag(negotiation_state()) :: non_neg_integer()
  def negotiation_state_to_tag(val) when is_map_key(@negotiation_state_tags, val) do
    Map.fetch!(@negotiation_state_tags, val)
  end

  @doc """
  All `NegotiationState` variants in tag order.
  """
  @spec all_negotiation_states() :: [negotiation_state()]
  def all_negotiation_states, do: [:neg_idle, :proposed, :selected, :neg_failed]

end
